packages/playwright-test/src/reporters/azure.ts

```diff
@@ -0,0 +1,481 @@
+  private token: string = '';
+
+  public constructor(options: AzureReporterOptions = {} as AzureReporterOptions) {
+    this._options = options;
+
+    this.orgUrl = this._options.orgUrl;
+    this.projectName = this._options.projectName;
+    this.environment = this._options.environment || undefined;
+    this.planId = this._options.planId;
+    this.logging = (typeof this._options.logging === 'undefined') ? true : this._options.logging;
+    this.isDisabled = this._options.isDisabled || false;
+    this.testRunTitle = `${this.environment ? `[${this.environment}]:` : ''} ${this._options.testRunTitle || 'Playwright Test Run'}` ||
+      `${this.environment ? `[${this.environment}]:` : ''}Test plan ${this.planId}`;
+    this.uploadAttachments = this._options.uploadAttachments || false;
+    this.attachmentsType = this._options.attachmentsType;
+    this.token = this._options.token;
+
+    if (!this.orgUrl || this.orgUrl.length === 0) {
+      this.log(colors.yellow("'orgUrl' is not set. Reporting is disabled."));
+      this.isDisabled = true;
+      return;
+    }
+    if (!this.projectName || this.projectName.length === 0) {
+      this.log(colors.yellow("'projectName' is not set. Reporting is disabled."));
+      this.isDisabled = true;
+      return;
+    }
+    if (!this.planId || this.planId === 0) {
+      this.log(colors.yellow("'planId' is not set. Reporting is disabled."));
+      this.isDisabled = true;
+      return;
+    }
+    if (!this.token || this.token.length === 0) {
+      this.log(colors.yellow("'token' is not set. Reporting is disabled."));
+      this.isDisabled = true;
+    }
+    if (this.uploadAttachments) {
+      if (!this.attachmentsType) {
+        this.log(colors.yellow("'attachmentsType' is not set. Attachments Type will be set to 'screenshot' by default."));
+        this.attachmentsType = ['screenshot'];
+      }
+    }
+    this.connection = new azdev.WebApi(this.orgUrl, azdev.getPersonalAccessTokenHandler(this.token));
+  }
+
+  async onBegin(): Promise<void> {
+    if (this.isDisabled)
+      return;
+
+    this.testApi = await this.connection.getTestApi();
+
+    this.createRun(this.testRunTitle).then(run => {
+      if (run) {
+        this.runId = run.id;
+        this.log(colors.green(`Using run ${this.runId} to publish test results`));
+      } else {
+        this.isDisabled = true;
+      }
+    });
+  }
+
+  async onTestEnd(test: TestCase, testResult: TestResult): Promise<void> {
+    await this.awaitForRunId().catch(e => {
+      this.log(colors.red(e));
+    });
+    if (this.isDisabled)
+      return;
+
+    this.logTestItem(test, testResult);
+    await this.publishCaseResult(test, testResult);
+  }
+
+  async onEnd(): Promise<void> {
+    await this.awaitForRunId().catch(e => {
+      this.log(colors.red(e));
+    });
+    if (this.isDisabled)
+      return;
+
+    let prevCount = this.resultsToBePublished.length;
+    while (this.resultsToBePublished.length > 0) {
```

Comment:
instead of having this logic here, I'd collect all the stuff in onTestEnd and then upload all the stuff on onEnd. So you don't need to wait for uploads to happen etc. and it makes it easier to understand.

Reply:
let's i'm trying to explain why i'm using this
1. if i'm upload results in `onTestEnd` user does not affected and results published parallel with main process, includes all attachments and the user will not even see how the test results are loaded.  If i'm try collect results and publish in `onEnd` - it would be have extra time. if it be only test results - it does not matter but if it includes attachments - it will be much more time to upload, because it's not local logic - it's network activity.
2. `resultsToBePublished` and `awaitForRunId`. if you test scope is more then `expect(1).toBe(1)`  - it does not matter. But if you have case like `expect(1).toBe(1); `- you need to await network activities like `createRun`, `publishCaseResult` because `onTestEnd` and `onEnd` hooks trigger much earlear and faster then network methods would be receive results. And we will receive result when main test run process will be ended but network requests not yet completed. If i'm remove this logic reporter will not work as expected of him, he will not have time to publish the results